### ISSUE TITLE
No unconstrained symbolic variables should ever be introduced to the path condition with region instantiation

### ISSUE THREAD

**Issue by vaibhavbsharma** [sohah/VeritestingTransformations#17]
Here's some code derived from TCAS that is failing a equivalence-check

```Java
public static int Cur_Vertical_Sep;
    public static int UNRESOLVED = 0;
    public static int MAXALTDIFF = 300;

    public int alt_sep_test() {
        int alt_sep = UNRESOLVED;
        if(Cur_Vertical_Sep > MAXALTDIFF){
            alt_sep = UNRESOLVED; // alt_assign();
        }
        return alt_sep;
    }
```

The summary we currently construct for this region is:
```
@w5$2 := 300; 
@w4$2 := in0; 
@w6$2 := 0; 
@w7$2 := (Gamma !(@w4$2<=@w5$2) @w6$2 @w3$2); 
```

Here's its CFG: 
[alt_sep_test_small.pdf](https://github.com/sohah/VeritestingTransformations/files/2443416/alt_sep_test_small.pdf)

Now the problem is: the gamma at the meet point uses `v3` but the region summary never includes `v3`'s definition. This causes the gamma to use a fresh symbolic variable (`w3$2` in Ranger IR) for `v3`. Here's what I think should happen:
1. We should identify all such variables and pull their defining statements into the region summary, if we can find them. 
2. We should never be instantiating a region with fresh unconstrained symbolic variables.

**Comment by vaibhavbsharma** [comment]
@sohah Another way of fixing this issue is to ensure that the region identifies `@w3` as an input to the region. It currently doesn't because of the way inputs are identified in ExprRegionInputVisitor. Can we change that design  to figure out that `@w3` should also be a input to the region ?

**Comment by sohah** [comment]
I am not sure Vaibhav, an input of the region is used by substitution to substitute its value either symbolic or concrete. A region input is defined by the first variable "use" for each stack slot inside the region. Now in this case, the what you have is actually a variable that is def, and we do not know anything about it for the substitution to handle it if we were to say that it is an input. 

I am not sure how this would fix the problem without actually grading the definition of the variable into the region, although I am not sure about the consequences of doing this either.

**Comment by vaibhavbsharma** [comment]
Yep, a region input will be the first "use" with the smallest Wala variable number for each stack slot. I've implemented this fix in [this](https://github.com/sohah/VeritestingTransformations/commit/f65e350457c62b004763cad249a27a34caa724ea) commit and it seems to work. It passes all of our small existing equivalence checks. It also passes equivalence checks in WBS (for 2 steps) and TCAS (for 1 step).

I think we should also implement a invariant that all Wala variables used in a region should either be defined in the same region or be in the local inputs table.

**Comment by vaibhavbsharma** [comment]
Yep, Soha and I looked over the code carefully. We both think that the change made to ExprRegionInputVisitor in [this](https://github.com/sohah/VeritestingTransformations/commit/f65e350457c62b004763cad249a27a34caa724ea) commit should be ok. I've also tested it for equivalence with WBS, TCAS, and other equivalence checks involving fields, arrays, and complex conditions. 

Having said that, this isn't a change that I was most confident on and I'll be keeping an eye on dangling unconstrained inputs in instantiated region summaries.
